### PR TITLE
fix: call getTABO when terminal is in view and expose new setting to set refetch interval

### PR DESCRIPTION
**Pull request by nghanyi**
- call `getTABO` only when terminal is in view (modal and widget)
- exposed a new property to let users set rpc refetch interval
- added setting to config form (default is set to 10000)
- update code snippet when form input changes
- fix npm code snippet not updating when config changed

image

image

### CODE REVIEW

File: src/constants/index.ts [hunk 1]
```diff
@@ -51,10 +52,11 @@ export const INITIAL_FORM_CONFIG: IFormConfigurator = Object.freeze({
...
+  rpcRefetchInterval: 10000,
 });
 
 export const TOKEN_2022_PROGRAM_ID = new PublicKey('TokenzQdBNbLqP5VEhdkAS6EPFLC1PHnBqCXEpPxuEb');
 
 export const DCA_HIGH_PRICE_IMPACT = 0.5; // 0.5%
 export const FREEZE_AUTHORITY_LINK =
-  'https://station.jup.ag/guides/jupiter-swap/how-swap-works/how-swap-works#freeze-authority';
\ No newline at end of file
+  'https://station.jup.ag/guides/jupiter-swap/how-swap-works/how-swap-works#freeze-authority';
```
Comment (nghanyi) [inline src/constants/index.ts:62]: didn't change this. think is just my formatter

Comment (nghanyi) [inline src/constants/index.ts:62]: didn't change this. think is just my formatter

File: src/components/FormConfigurator.tsx [hunk 2]
```diff
@@ -106,10 +107,11 @@ const FormConfigurator = ({
   const { query, replace } = useRouter();
 
   const [isImported, setIsImported] = useState(false);
+  const [configureRpcRefetchInterval, setConfigureRpcRefetchInterval] = useState(rpcRefetchInterval !== undefined);
 
   const onSelect = useCallback(
     (index: number) => {
-      console.log('called')
+      console.log('called');
```
Comment (nghanyi) [inline src/components/FormConfigurator.tsx:114]: auto formatter

Comment (gabrielkhh) [reply src/components/FormConfigurator.tsx:114]: imo we can remove these console logs though, doesnt seem necessary to me to leave it there

File: src/types/index.d.ts [hunk 3]
```diff
@@ -92,6 +92,8 @@ export interface IInit {
   useUserSlippage?: boolean;
   /** TODO: NOT Supported yet, presets of slippages, defaults to [0.1, 0.5, 1.0] */
   slippagePresets?: number[];
+  /** RPC refetch interval for getTABO in milliseconds, defaults to 10000 */
+  rpcRefetchInterval?: number;
```
Comment (worlddlckgh) [inline src/types/index.d.ts:96]: refetchIntervalForTokenAccounts?: number;

File: src/components/FormConfigurator.tsx [hunk 4]
```diff
@@ -421,6 +423,40 @@ const FormConfigurator = ({
...
+            setConfigureRpcRefetchInterval(!configureRpcRefetchInterval);
+          }}
+        />
+      </div>
+      {configureRpcRefetchInterval && (
+        <input
+          className="mt-2 text-white w-full flex justify-between items-center space-x-2 text-left rounded-md bg-white/10 px-4 py-2 text-sm font-medium shadow-sm border border-white/10"
+          value={rpcRefetchInterval}
+          inputMode="numeric"
+          placeholder="10000"
```
Comment (worlddlckgh) [inline src/components/FormConfigurator.tsx:449]: `10000 (10s)`
so it's easier to understand

File: src/components/FormConfigurator.tsx [hunk 5]
```diff
@@ -106,10 +107,12 @@ const FormConfigurator = ({
   const { query, replace } = useRouter();
 
   const [isImported, setIsImported] = useState(false);
+  const [configureRefetchIntervalForTokenAccounts, setConfigureRefetchIntervalForTokenAccounts] = useState(
```
Comment (wkitsin) [inline src/components/FormConfigurator.tsx:110]: actually we dont need this right ? just use the form value ? and set it to undefined if we toggle it off

Comment (nghanyi) [reply src/components/FormConfigurator.tsx]: good point. i have removed this `useState` and use the one in form value

File: src/contexts/accounts.tsx [hunk 6]
```diff
@@ -203,8 +208,8 @@ const AccountsProvider: React.FC<PropsWithChildren> = ({ children }) => {
       };
     },
     {
-      enabled: Boolean(publicKey?.toString() && connected),
-      refetchInterval: 10_000,
+      enabled: Boolean(publicKey?.toString() && connected && getTerminalInView()),
+      refetchInterval: refetchIntervalForTokenAccounts,
```
Comment (worlddlckgh) [inline src/contexts/accounts.tsx:212]: we should still keep 10s as default

File: src/constants/index.ts [hunk 7]
```diff
@@ -51,6 +52,7 @@ export const INITIAL_FORM_CONFIG: IFormConfigurator = Object.freeze({
     initialOutputMint: WRAPPED_SOL_MINT.toString(),
   },
   useUserSlippage: true,
+  refetchIntervalForTokenAccounts: 10000,
```
Comment (worlddlckgh) [inline src/constants/index.ts:55]: _app.tsx and FormConfigurator are intended for Terminal preview site.
you should have 10_000 passed down from somewhere, or directly in the accounts.tsx contexts so the default behaviour is kept

File: src/library.tsx [hunk 8]
```diff
@@ -198,7 +203,10 @@ const RenderWidgetShell = (props: IInit) => {
     <div className={`fixed ${classes.containerClassName}`}>
       <div
         className={`${classes.widgetContainerClassName} rounded-full bg-black flex items-center justify-center cursor-pointer`}
-        onClick={() => setIsOpen(!isOpen)}
+        onClick={() => {
+          setIsOpen(!isOpen);
+          setTerminalInView(!getTerminalInView());
+        }}
```
Comment (worlddlckgh) [inline src/library.tsx:209]: if `isOpen` and `getTerminalInView()` is not in sync, you will end up in crazy state toggling that never matches up
you should rely on `isOpen` state as the source of truth, so it stays in sync

```
() => {
  if (isOpen) {
    setIsOpen(false);
    setTerminalInView(false);
  } else {
     // ... true true
  }
 
}
```

File: src/contexts/accounts.tsx (hunk 6 above)
Comment (nghanyi) [reply src/contexts/accounts.tsx:212]: have set default value of `refetchIntervalForTokenAccounts` to 10s

File: src/constants/index.ts (hunk 7 above)
Comment (nghanyi) [reply src/constants/index.ts]: removed from form

**Comment by worlddlckgh** [comment]
Taking over this, thanks for the contrib!

**Pull request merged**